lib/elixir/lib/system.ex

```diff
@@ -736,18 +736,25 @@ defmodule System do
   @doc """
   Sets an environment variable value.
 
-  Sets a new `value` for the environment variable `varname`.
+  Sets a new `value` for the environment variable `key`.
+  `nil` values erase the given key.
   """
-  @spec put_env(binary, binary) :: :ok
-  def put_env(varname, value) when is_binary(varname) and is_binary(value) do
-    case :binary.match(varname, "=") do
-      {_, _} ->
-        raise ArgumentError,
-              "cannot execute System.put_env/2 for key with \"=\", got: #{inspect(varname)}"
+  @spec put_env(binary, binary | nil) :: :ok
+  def put_env(key, value) when is_binary(key) do
+    case value do
```

Comment:
Let's use pattern matching on the function had, it will require fewer changes to the codebase:

```
  @spec put_env(binary, binary | nil) :: :ok
  def put_env(varname, nil) when is_binary(varname), do: ...

  def put_env(...) do
    ...current code...
  end
```